components/lib/nearby/src/main/java/mozilla/components/lib/nearby/NearbyConnection.kt

```diff
@@ -0,0 +1,313 @@
+ * called from the UI thread or externally synchronized.
+ *
+ * @constructor Constructs a new connection, which will call [NearbyConnectionListener.updateState]
+ *     with an argument of type [ConnectionState.Isolated]. No further action will be taken unless
+ *     other methods are called by the client.
+ * @param context context needed to initiate connection, used only at start
+ * @param name name shown by this device to other devices
+ * @param authenticate whether to authenticate the connection (true) or make it automatically (false)
+ * @param listener listener to be notified of changes of state and message transmission
+ *
+ */
+@UiThread
+class NearbyConnection(
+    private val context: Context,
+    private val name: String,
+    private val authenticate: Boolean,
+    private val listener: NearbyConnectionListener
+) {
+    // Compile-time constants
+    private val TAG = "NearbyConnection"
+    private val PACKAGE_NAME = "mozilla.components.feature.nearby.NearbyConnection"
+    private val STRATEGY = Strategy.P2P_STAR
+
+    /**
+     * The state of the connection. Changes in state are communicated through
+     * [NearbyConnectionListener.updateState].
+     *
+     * @property name: an English-language name for the state
+     */
+    public sealed class ConnectionState(val name: String) {
+        object Isolated : ConnectionState("isolated")
+        object Advertising : ConnectionState("advertising")
+        object Discovering : ConnectionState("discovering")
+        class Authenticating(
+            // sealed classes can't be inner, so we need to pass in the connection
+            private val nearbyConnection: NearbyConnection,
+            val neighborId: String,
+            val neighborName: String,
+            val token: String
+        ) : ConnectionState("authenticating") {
+            /**
+             * Prompts the user to accept or reject the connection to [neighborName] with the
+             * given [token]. Specifically, this shows an [androidx.appcompat.app.AlertDialog] with
+             * the given [title] and [message] and calls [accept] or [reject] based on the user's
+             * choice. This method is provided for convenience. Clients can implement their own
+             * dialog or logic.
+             *
+             * @param context the context for the AlertDialog
+             * @param title the title of the AlertDialog
+             * @param message the message in the AlertDialog
+             */
+            fun showAuthenticationDialog(
+                context: Context,
+                title: String = "Accept connection to $neighborName",
+                message: String = "Confirm the code matches on both devices: $token"
+            ) {
+                AlertDialog.Builder(context)
+                    .setTitle(title)
+                    .setMessage(message)
+                    .setPositiveButton(
+                        "Accept") { _, _ ->
```

Comment:
Yeah, I know better. I'll do that now.

Reply:
I see the informative comments that Mozillians put before the strings. That makes me very happy. Usually I'm the one reminding people of proper practices.

Reply:
Would it be appropriate for me to use `android.R.string.yes` in this case?

Reply:
I'm willing to add a string resources file to the library, but I feel a little uncomfortable mixing my view and my model. Of course, I already started down this path by putting an `AlertDialog` in the library. Do you think I should (1) add string resources, (2) require the client to pass in the strings, (3) leave creating a dialog to the client, or (4) something else?

Reply:
I think 3 would be most optimal (I mentioned this in another comment as well), since not all users would want to show the same form of confirmation (i.e. maybe a dialog/toast/unknown option).

> Would it be appropriate for me to use `android.R.string.yes` in this case?

`android.R.string.yes` ,I think that would be fine. 🥁 😄

Reply:
> I see the informative comments that Mozillians put before the strings. That makes me very happy. Usually I'm the one reminding people of proper practices.

Our string files are collected together by L10N that send them off to get translated, and the comment on each of them get sent along as context for the translator. There's a check somewhere as well to make sure there are always comments. :)

Reply:
I removed the method.